src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -481,14 +482,18 @@ private NamedArgsAndClosure(Map<?,?> namedArgs, Closure body) {
      * but better to do it here in the Groovy-specific code so we do not need to rely on that.
      * @return {@code v} or an equivalent with all {@link GString}s flattened, including in nested {@link List}s or {@link Map}s
      */
-    private static Object flattenGString(Object v) {
+    private static Object flattenGString(Object v, @Nullable EnvironmentWatcher envWatcher) {
         if (v instanceof GString) {
-            return v.toString();
+            String flattened = v.toString();
+            if (envWatcher != null) {
+                envWatcher.scan(flattened);
```

Comment:
I actually hadn't considered that scenario, but the fact that  it can create a false negative makes a bit wary to scan only in  `getValues`.  I'm not too worried  about false  positives because  I think they would only really trigger if someone  had a  really bad password like `String` (please  correct me if  I'm wrong!)